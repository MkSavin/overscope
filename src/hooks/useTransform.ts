import { useMemo, useRef } from 'react'
import { produce } from 'immer'
import {
  OverscopePatch, OverscopeSetter, OverscopeState, OverscopeTransform, OverscopeTransformer,
} from '../types'

/**
 * Create overscope transform object.
 *
 * @param stateHandle state setter or [ state, setter ].
 *   If `state` provided it used for initial value of `stateRef`.
 * @param transformer User-defined transform object
 */
export const useTransform = (
  <
    State extends OverscopeState,
    Transform extends OverscopeTransform,
  >(
    stateHandle: OverscopeSetter<State> | [ State, OverscopeSetter<State> ],
    transformer: OverscopeTransformer<State, Transform>,
  ): Transform => {
    const state = Array.isArray(stateHandle)
      ? stateHandle[0]
      : undefined

    const stateRef = useRef(state)
    stateRef.current = state

    const setter = Array.isArray(stateHandle)
      ? stateHandle[1]
      : stateHandle

    const setterRef = useRef(setter)
    setterRef.current = setter

    const patch: OverscopePatch<State> = (callback) => {
      setterRef.current((current) => {
        const modified = produce(current, callback)
        stateRef.current = modified
        return modified
      })
    }

    return useMemo(() => {
      const bareTransform = transformer(patch, stateRef)

      const beforeCallback = bareTransform._beforeEach
      const afterCallback = bareTransform._afterEach

      const richTransform = Object.entries(bareTransform)
        .filter(([ code ]) => code !== '_beforeEach' && code !== '_afterEach')
        .map(([ code, handle ]) => {
          if (typeof handle !== 'function') {
            return [ code, handle ]
          }

          return [
            code,
            (...data: any[]): void => {
              beforeCallback?.()
              handle(...data)
              afterCallback?.()
            },
          ]
        })

      return Object.fromEntries(richTransform)
    }, [])
  }
)
