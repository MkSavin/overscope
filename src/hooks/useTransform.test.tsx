import React, { FC, useEffect, useState } from 'react'
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react'
import { useTransform } from './useTransform'

type TestState = {
  count: number,
}

type TestTransform = {
  increase: () => void,
  superFunction?: (number: number) => void,
}

type TestNestedTransform = {
  nested: {
    increase: () => void,
  },
}

describe('useTransform', () => {
  it('Modifies provided store', () => {
    const TestComponent: FC = () => {
      const [ state, setState ] = useState<TestState>({
        count: 0,
      })

      const transform = useTransform<TestState, TestTransform>(setState, (patch) => ({
        increase: () => patch((current) => {
          current.count++
        }),
      }))

      useEffect(() => {
        transform.increase()
      }, [])

      return (
        <>{ state.count }</>
      )
    }

    const { asFragment } = render((
      <TestComponent />
    ))

    expect(asFragment()).toMatchSnapshot()
  })

  it('Provides actual state reference', () => {
    const TestComponent: FC = () => {
      const [ state, setState ] = useState<TestState>({
        count: 1,
      })

      const transform = useTransform<TestState, TestTransform>(
        [ state, setState ],
        (patch, stateRef) => ({
          increase: () => {
            const outPatch = stateRef.current?.count ?? 0

            patch((current) => {
              current.count = outPatch + 1
            })
          },
        }),
      )

      useEffect(() => {
        transform.increase()
      }, [])

      return (
        <>{ state.count }</>
      )
    }

    const { asFragment } = render((
      <TestComponent />
    ))

    expect(asFragment()).toMatchSnapshot()
  })

  it('Executes event handlers', () => {
    const handleEvent = jest.fn()

    const TestComponent: FC = () => {
      const [ state, setState ] = useState<TestState>({
        count: 1,
      })

      const transform = useTransform<TestState, TestTransform>(
        [ state, setState ],
        (patch, stateRef) => ({
          increase: () => {
            const outPatch = stateRef.current?.count ?? 0

            patch((current) => {
              current.count = outPatch + 1
            })
          },

          _beforeEach: () => {
            const outPatch = stateRef.current?.count ?? 0

            handleEvent()

            patch((current) => {
              current.count = outPatch + 10
            })
          },

          _afterEach: () => {
            const outPatch = stateRef.current?.count ?? 0

            handleEvent()

            patch((current) => {
              current.count = outPatch * 2
            })
          },
        }),
      )

      useEffect(() => {
        transform.increase()
      }, [])

      return (
        <>{ state.count }</>
      )
    }

    const { asFragment } = render((
      <TestComponent />
    ))

    expect(asFragment()).toMatchSnapshot()

    expect(handleEvent).toBeCalledTimes(2)
  })

  it('Do not executes event handlers for nested callbacks', () => {
    const handleEvent = jest.fn()

    const TestComponent: FC = () => {
      const [ state, setState ] = useState<TestState>({
        count: 1,
      })

      const transform = useTransform<TestState, TestNestedTransform>(
        [ state, setState ],
        (patch, stateRef) => ({
          nested: {
            increase: () => {
              const outPatch = stateRef.current?.count ?? 0

              patch((current) => {
                current.count = outPatch + 1
              })
            },
          },

          _beforeEach: () => {
            const outPatch = stateRef.current?.count ?? 0

            handleEvent()

            patch((current) => {
              current.count = outPatch + 10
            })
          },

          _afterEach: () => {
            const outPatch = stateRef.current?.count ?? 0

            handleEvent()

            patch((current) => {
              current.count = outPatch * 2
            })
          },
        }),
      )

      useEffect(() => {
        transform.nested.increase()
      }, [])

      return (
        <>{ state.count }</>
      )
    }

    const { asFragment } = render((
      <TestComponent />
    ))

    expect(asFragment()).toMatchSnapshot()

    expect(handleEvent).toBeCalledTimes(0)
  })
})
