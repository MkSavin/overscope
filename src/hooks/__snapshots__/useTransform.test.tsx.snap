// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTransform Do not executes event handlers for nested callbacks 1`] = `
<DocumentFragment>
  2
</DocumentFragment>
`;

exports[`useTransform Executes event handlers 1`] = `
<DocumentFragment>
  22
</DocumentFragment>
`;

exports[`useTransform Modifies provided store 1`] = `
<DocumentFragment>
  1
</DocumentFragment>
`;

exports[`useTransform Provides actual state reference 1`] = `
<DocumentFragment>
  2
</DocumentFragment>
`;
